### DISCUSSION TITLE
Lesson 7 : brownie.exceptions.VirtualMachineError: revert

### DISCUSSION THREAD

**Discussion by Krishduttm** [smartcontractkit/full-blockchain-solidity-course-py#970] [Q&A] [answered]
```
Brownie v1.17.2 - Python development framework for Ethereum

==================================================== test session starts ====================================================
platform darwin -- Python 3.9.5, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/apple/Desktop/web3/Brownie/brownie-smart-lottery_
plugins: eth-brownie-1.17.2, xdist-1.34.0, hypothesis-6.27.3, web3-5.25.0, forked-1.3.0
collected 1 item                                                                                                            

Launching 'ganache-cli --accounts 10 --fork https://eth-rinkeby.alchemyapi.io/v2/OzP_8l22Ebcxi4UKfbsX40aGnCcfPD65 --mnemonic brownie --port 8545 --hardfork istanbul'...

tests/test_lottery.py F                                                                                               [100%]

========================================================= FAILURES ==========================================================
_________________________________________________ test_Lottery_entranceFee __________________________________________________

    def test_Lottery_entranceFee():
    
        account = accounts[0]
        lottery = Lottery.deploy(
            config["networks"][network.show_active()]["eth_usd_price_feed"],
            {"from": account},
        )
>       assert lottery.getEntranceFee() < 2000000000000000
E       brownie.exceptions.VirtualMachineError: revert

tests/test_lottery.py:12: VirtualMachineError
================================================== short test summary info ==================================================
FAILED tests/test_lottery.py::test_Lottery_entranceFee - brownie.exceptions.VirtualMachineError: revert
===================================================== 1 failed in 7.07s =====================================================
```
<img width="1440" alt="Screenshot 2022-02-07 at 12 04 18 AM" src="https://user-images.githubusercontent.com/59798012/152695989-1235690d-60cc-4b68-a73d-081dfd936e72.png">

what is the problem?

My brownie-config.yaml - 
<img width="1440" alt="Screenshot 2022-02-06 at 5 08 43 PM" src="https://user-images.githubusercontent.com/59798012/152679087-357d3af3-63b4-4e23-bb4c-38aab49bc4a4.png">

test-file - 
<img width="1440" alt="Screenshot 2022-02-06 at 5 10 57 PM" src="https://user-images.githubusercontent.com/59798012/152679166-cb5333dd-c948-4d91-83a7-cd953bcfb9af.png">

contract file - 
<img width="1440" alt="Screenshot 2022-02-06 at 5 11 34 PM" src="https://user-images.githubusercontent.com/59798012/152679195-07d1d73d-b15b-420e-9bcf-e175d1c7cfe0.png">

**Comment by cromewar** [comment]
Hello @Krishduttm 
Could you run the test again but using -s identifier.

`brownie test -s -k tes_name --network network_name`

**Reply by sheahwen** [reply]
Hi @cromewar
I have the exact same issue as @Krishduttm. Same test function and setup. I have ran the test again using
`brownie test -s -k test_get_entrance_fee --network mainnet-fork`

<img width="824" alt="error1" src="https://user-images.githubusercontent.com/25454526/154836882-cf525004-0ddb-404a-852c-5f72ab092b12.PNG">
<img width="658" alt="error2" src="https://user-images.githubusercontent.com/25454526/154836883-cce9913b-73be-497c-b31d-1ba24dd5fbe7.PNG">

I am currently following Patrick's code till 6:29:27. Appreciate if you could help me with the issue. Thanks!

**Reply by Krishduttm** [reply]
Can you share your test file code and contract file? Thanks

**Reply by sheahwen** [reply]
Following are my codes. Brownie compile works perfectly. I have upgraded to brownie version18 as suggested by @n4n0b1t3 but still having the same error. 

Lottery.sol
```Solidity
// SPDX-License-Identifier: MIT

pragma solidity >=0.6.0 <0.9.0;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";

contract Lottery {
    address payable[] public players;
    uint256 public usdEntryFee;
    AggregatorV3Interface internal ethUsdPriceFeed;

    constructor(address _priceFeedAddress) public {
        usdEntryFee = 50 * (10**18);
        ethUsdPriceFeed = AggregatorV3Interface(_priceFeedAddress);
    }
    
    function enter() public payable {
        // minimum USD$50
        // require();
        players.push(msg.sender);
    }
    function getEntranceFee() public view returns (uint256) {
        (,int256 price,,,) = ethUsdPriceFeed.latestRoundData();
        // $50 => $50/ETH
        uint256 adjustedPrice = uint256(price) * 10 **10; // 18 decimals
        uint256 costToEnter = (usdEntryFee * 10 ** 18) / adjustedPrice;
        return costToEnter;
    }
    function startLottery() public {}
    function endLottery() public {}
}
```

test_lottery.py
```Python
from brownie import Lottery, accounts, config, network


def test_get_entrance_fee():
    account = accounts[0]
    lottery = Lottery.deploy(
        config["networks"][network.show_active()]["eth_usd_price_feed"],
        {"from": account},
    )
    assert lottery.getEntranceFee() > 1000
```

brownie-config.yaml
```YAML
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
compiler:
  solc:
    remappings:
      - "@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1"
networks:
  mainnet-fork:
    eth_usd_price_feed: "0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419"
```

**Reply by cromewar** [reply]
For me the problem is due to a bad configuration on the network
Please delete mainnet-fork and add it again, being sure to not have a type and typing correctly the host name

**Comment by n4n0b1t3** [comment, marked as answer]
when walking backward from your error through your code I come to only one transaction which can revert.
![image](https://user-images.githubusercontent.com/32838961/154837803-d2dcba71-a020-4d16-ba13-3e684dc014a9.png)

Since you are using the mainnet-fork, try to dissect your code by opening a new *.py file and only use the minimum code necessary to test the correct workings of the AggregatorV3Interface(_priceFeedAddress) transaction. 

I also noticed that you have an outdated Brownie version. The current version is 1.18.1.
Please look at: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1110
and: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1109

The whole setup might be a little cumbersome the first time, but you can eliminate a lot of potential errors which have nothing to do with your code, by just implementing this approach.

**Reply by Krishduttm** [reply]
Hey , thanks for the reply , the problem was that i tried testing for different scenarios based on the current price of ethereum. I worked when i gave whatever patrick did in the course. I actually still have this doubt as to why the pricefeed didnt give the current price of ethereum. If you can help me on that thanks!

**Reply by n4n0b1t3** [reply]
If you have your Infura ID setup in your .env file you can check what chainlink will return from the mainnet-fork.

```python
brownie console --network mainnet-fork
Brownie environment is ready.
>>> address = "0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419"
>>> oracle = Contract.from_explorer(address)
>>> oracle.latestAnswer()
265282750767
>>>
```

**Reply by cromewar** [reply]
Did @n4n0b1t3 solution worked for you @Krishduttm ?

**Reply by Krishduttm** [reply]
<img width="1440" alt="Screenshot 2022-02-22 at 12 01 22 PM" src="https://user-images.githubusercontent.com/59798012/155075590-9e6558c2-8835-4fed-b738-034b5a5078f3.png">


getting this error. I have my infura set up in .env

**Reply by cromewar** [reply]
@Krishduttm try using `export` on your console and export the infura key before running the deploy/test. Sometimes the .env is not read correctly

**Reply by dRoysch** [reply]
Hi! I have the same issue I am working with Alchemy just like [6:28](https://www.youtube.com/watch?v=M576WGiDBdQ&ab_channel=freeCodeCamp.org)

In the test_lottery.py I have the error with
 
`lotttery.getEntranceFee()`
And if I print(lottery.getEntranceFee) this returns:
`<ContractCall 'getEntranceFee()'>`

I am also check de constructor of Lottery.sol and is exact what @Krishduttm  pressents above.

In the brownie console I have the same issue that him.

But with alchemy I understand that we only need this config in the brownie network
`brownie networks add development mainnet-fork cmd=ganache-cli host=http://127.0.0.1 fork=wss://eth-mainnet.alchemyapi.io/v2/my_api  accounts=10 mnemonic=brownie port=8545`



I will appreciate your help, I am spending a lot of time trying to fix this little bug :(

**Reply by jslim89** [reply]
Hi, I'm having the same issue, has been stuck for days

Python: v3.8
Brownie v1.18.1
solidity: ^0.8.0;
smartcontractkit/chainlink-brownie-contracts@1.2.0

<img width="1434" alt="image" src="https://user-images.githubusercontent.com/797182/162614282-0f9a1763-e394-4924-b753-4e28d1ca4398.png">

I've tried both Alchemy & Infura, tried Python 3.8 & 3.9, tried Solidity 0.6.6 & 0.8.0, tried both `http` & `wss` in the fork, all not working

```
brownie networks add Development mainnet-fork cmd=ganache-cli host=http://ganache-cli fork=https://mainnet.infura.io/v3/apikey accounts=10 mnemonic=brownie port=8545
brownie networks add Development mainnet-fork cmd=ganache-cli host=http://ganache-cli fork='wss://eth-mainnet.alchemyapi.io/v2/apikey' accounts=10 mnemonic=brownie port=8545
```

Anything else I might possibly miss?

**Reply by cromewar** [reply]
Did you guys @jslim89 @dRoysch deleted the default mainnet-fork network before adding the new one? also have you configured the infura/alchemy on the correct networks?

**Reply by jslim89** [reply]
Hi @cromewar Yes, I have deleted the existing mainnet-fork, otherwise it will not be allowed to add another one.

The Alchemy config

<img width="704" alt="image" src="https://user-images.githubusercontent.com/797182/162651162-526b9452-3f51-446d-aa10-ae9bc393bc6b.png">

**Reply by dRoysch** [reply]
Hi!

I fix the problem by closing Ganache. 

I saw the error because when I `ganache-cli` into the console
Throw this error
`Ganache CLI v6.12.2 (ganache-core: 2.13.2)
Error: listen EADDRINUSE: address already in use 127.0.0.1:8545
    at Server.setupListenHandle [as _listen2] (net.js:1301:14)
    at listenInCluster (net.js:1349:12)
    at doListen (net.js:1488:7)
    at processTicksAndRejections (internal/process/task_queues.js:81:21)`

Note: when command `brownie test --network mainnet-fork` the contract was deployed to Ganache GUI.

Thanks you a lot!

**Reply by cromewar** [reply]
I'm glad is solved now.
